google/cloud/forseti/scanner/scanners/external_project_access_scanner.py

```diff
@@ -0,0 +1,197 @@
+                'violation_type': violation.violation_type,
+                'violation_data': violation_data,
+                'resource_data': violation.resource_data
+            }
+
+    def _retrieve(self):
+        # pylint: disable=line-too-long
+        """Retrieve the project ancestries for all users.
+
+        Returns:
+            dict: User project relationship.
+            {"user1@example.com": [[Project("1234"), Organization("1234567")],
+                                  [Project("12345"), Folder("ABCDEFG"), Organization("1234567")]],
+             "user2@example.com": [[Project("1234"), Organization("1234567")],
+                                  [Project("12345"), Folder("ABCDEFG"), Organization("1234567")]]}
+        """
+        # pylint: enable=line-too-long
+        # This dictionary is the result of the scan.  The key
+        # is the user ID.  The value is a list of lists of ancestries.
+        project_ancestries_by_user = {}
+        user_count = 0
+
+        start_time = time.time()
+
+        user_emails = api_data_helpers.get_user_emails(self.service_config)
+
+        for user_email in user_emails:
+            user_count += 1
+
+            user_scoped_credential = \
+                api_helpers.get_delegated_credential(user_email, SCOPES)
+
+            client = CloudResourceManagerClient(
+                global_configs=self.inventory_configs,
+                credentials=user_scoped_credential)
+
+            try:
+                project_response = client.get_projects()
+
+                project_ids = api_data_helpers.get_project_ids(project_response)
+                for project_id in project_ids:
+                    if project_id not in DISCOVERED_ANCESTRIES.keys():
+                        ancestries = \
+                            client.get_project_ancestry(project_id)
+                        ancestry_resources = \
+                            api_data_helpers.get_project_ancestry_resources(
+                                ancestries)
+                        DISCOVERED_ANCESTRIES[project_id] = ancestry_resources
+
+                    if user_email not in project_ancestries_by_user.keys():
+                        project_ancestries_by_user[user_email] = []
+
+                    project_ancestries_by_user[user_email].append(
+                        DISCOVERED_ANCESTRIES[project_id])
+            except KeyError:
+                # Occurs if the user is not a member of any projects.
+                LOGGER.debug('User %s doesn\'t have any projects.',
+                             user_email)
+
+        elapsed_time = time.time() - start_time
+
+        LOGGER.debug('It took %f seconds to query projects for %d users',
+                     elapsed_time,
+                     user_count)
+
+        return project_ancestries_by_user
+
+    def run(self):
+        """Entry point to run the scanner."""
+        project_ancestries_by_user = self._retrieve()
```

Comment:
As an outsider, the naming of this variable would be easier to read as `user_to_project_ancestries_map`, i.e. `key_to_value_map`.